File: src/main/java/raylras/zen/l10n/UTF8Control.java
package raylras.zen.l10n;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.Locale;
import java.util.PropertyResourceBundle;
import java.util.ResourceBundle;

// https://stacktuts.com/how-to-use-utf-8-in-resource-properties-with-resourcebundle
public class UTF8Control extends ResourceBundle.Control {

    @Override
    public ResourceBundle newBundle(String baseName, Locale locale, String format, ClassLoader loader, boolean reload) throws IOException {
        String bundleName = toBundleName(baseName, locale);
        String resourceName = toResourceName(bundleName, "properties");
        try (InputStream stream = loader.getResourceAsStream(resourceName)){
            if (stream != null) {
                // Read properties files as UTF-8
                return new PropertyResourceBundle(new InputStreamReader(stream, StandardCharsets.UTF_8));
            } else {
                return null;
            }
        }
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/util/l10n/UTF8Control.kt
package raylras.zen.util.l10n

import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.*

// https://stacktuts.com/how-to-use-utf-8-in-resource-properties-with-resourcebundle
class UTF8Control : ResourceBundle.Control() {
    override fun newBundle(
        baseName: String,
        locale: Locale,
        format: String,
        loader: ClassLoader,
        reload: Boolean
    ): ResourceBundle {
        val bundleName = toBundleName(baseName, locale)
        val resourceName = toResourceName(bundleName, "properties")
        loader.getResourceAsStream(resourceName).use { stream ->
            // Read properties files as UTF-8
            return PropertyResourceBundle(InputStreamReader(stream!!, StandardCharsets.UTF_8))
        }
    }
}
